Fix web android to behave the same as iOS

wojtus7 opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please follow the template so that the reviewers can easily understand what the code changes affect. -->

## Summary

<!-- Explain the motivation for this PR. Include "Fixes #<number>" if applicable. -->

## Test plan

<!-- Provide a minimal but complete code snippet that can be used to test out this change along with instructions how to run it and a description of the expected behavior. -->

wojtus7 committed:
Fix web android to behave the same as iOS

wojtus7 committed:
Merge remote-tracking branch 'origin/fix-gyroscope' into wojtus7/web-android-sensor

dependabot[bot] committed:
Bump semver from 5.7.1 to 5.7.2 in /docs (#4948)

Bumps [semver](https://github.com/npm/node-semver) from 5.7.1 to 5.7.2.
<details>
<summary>Release notes</summary>
<p><em>Sourced from <a
href="https://github.com/npm/node-semver/releases">semver's
releases</a>.</em></p>
<blockquote>
<h2>v5.7.2</h2>
<h2><a
href="https://github.com/npm/node-semver/compare/v5.7.1...v5.7.2">5.7.2</a>
(2023-07-10)</h2>
<h3>Bug Fixes</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/2f8fd41487acf380194579ecb6f8b1bbfe116be0"><code>2f8fd41</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/585">#585</a>
better handling of whitespace (<a
href="https://redirect.github.com/npm/node-semver/issues/585">#585</a>)
(<a href="https://github.com/joaomoreno"><code>@​joaomoreno</code></a>,
<a
href="https://github.com/lukekarrys"><code>@​lukekarrys</code></a>)</li>
</ul>
</blockquote>
</details>
<details>
<summary>Changelog</summary>
<p><em>Sourced from <a
href="https://github.com/npm/node-semver/blob/v5.7.2/CHANGELOG.md">semver's
changelog</a>.</em></p>
<blockquote>
<h2><a
href="https://github.com/npm/node-semver/compare/v5.7.1...v5.7.2">5.7.2</a>
(2023-07-10)</h2>
<h3>Bug Fixes</h3>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/2f8fd41487acf380194579ecb6f8b1bbfe116be0"><code>2f8fd41</code></a>
<a href="https://redirect.github.com/npm/node-semver/pull/585">#585</a>
better handling of whitespace (<a
href="https://redirect.github.com/npm/node-semver/issues/585">#585</a>)
(<a href="https://github.com/joaomoreno"><code>@​joaomoreno</code></a>,
<a
href="https://github.com/lukekarrys"><code>@​lukekarrys</code></a>)</li>
</ul>
<h2>5.7</h2>
<ul>
<li>Add <code>minVersion</code> method</li>
</ul>
<h2>5.6</h2>
<ul>
<li>Move boolean <code>loose</code> param to an options object, with
backwards-compatibility protection.</li>
<li>Add ability to opt out of special prerelease version handling with
the <code>includePrerelease</code> option flag.</li>
</ul>
<h2>5.5</h2>
<ul>
<li>Add version coercion capabilities</li>
</ul>
<h2>5.4</h2>
<ul>
<li>Add intersection checking</li>
</ul>
<h2>5.3</h2>
<ul>
<li>Add <code>minSatisfying</code> method</li>
</ul>
<h2>5.2</h2>
<ul>
<li>Add <code>prerelease(v)</code> that returns prerelease
components</li>
</ul>
<h2>5.1</h2>
<ul>
<li>Add Backus-Naur for ranges</li>
<li>Remove excessively cute inspection methods</li>
</ul>
<h2>5.0</h2>
<ul>
<li>Remove AMD/Browserified build artifacts</li>
<li>Fix ltr and gtr when using the <code>*</code> range</li>
<li>Fix for range <code>*</code> with a prerelease identifier</li>
</ul>
</blockquote>
</details>
<details>
<summary>Commits</summary>
<ul>
<li><a
href="https://github.com/npm/node-semver/commit/f8cc313550691a50d9662d8c94f0c033717efd7d"><code>f8cc313</code></a>
chore: release 5.7.2</li>
<li><a
href="https://github.com/npm/node-semver/commit/2f8fd41487acf380194579ecb6f8b1bbfe116be0"><code>2f8fd41</code></a>
fix: better handling of whitespace (<a
href="https://redirect.github.com/npm/node-semver/issues/585">#585</a>)</li>
<li><a
href="https://github.com/npm/node-semver/commit/deb5ad51bf58868fa243c1683775305fe9e0e365"><code>deb5ad5</code></a>
chore: <code>@​npmcli/template-oss</code><a
href="https://github.com/4"><code>@​4</code></a>.16.0</li>
<li>See full diff in <a
href="https://github.com/npm/node-semver/compare/v5.7.1...v5.7.2">compare
view</a></li>
</ul>
</details>
<details>
<summary>Maintainer changes</summary>
<p>This version was pushed to npm by <a
href="https://www.npmjs.com/~lukekarrys">lukekarrys</a>, a new releaser
for semver since your current version.</p>
</details>
<br />


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=semver&package-manager=npm_and_yarn&previous-version=5.7.1&new-version=5.7.2)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot show <dependency name> ignore conditions` will show all
of the ignore conditions of the specified dependency
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)
You can disable automated security fix PRs for this repo from the
[Security Alerts
page](https://github.com/software-mansion/react-native-reanimated/network/alerts).

</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

bartlomiejbloniarz committed:
Add eslint rule for consistent `import type` (#4950)

## Summary
This PR adds an eslint rule that forces `import type` to always have the
`type` specifier on the top level.

![Screenshot 2023-08-21 at 17 27
17](https://github.com/software-mansion/react-native-reanimated/assets/56109050/a328f3fd-23d4-47bb-a374-24316d7f1cdf)

## Test plan

<!-- Provide a minimal but complete code snippet that can be used to
test out this change along with instructions how to run it and a
description of the expected behavior. -->

michalmaka committed:
Better reanimated shadow tree commit detection (#4949)

## Summary

Proposed solution is more bullet proof solution for detecting whether
currently committed shadow tree comes from reanimated.

Here are the main issues with current implementation:
- if there's another commit hook that is executed before `reanimated`
one, then `propsRegistry_->isLastReanimatedRoot(newRootShadowNode)` may
not work if the commit hook creates a copy of committed tree
- when `stateReconciliation` is enabled via commit options then new tree
root is created and again above detection does not work.

It uses concept of `thread_local` c++ variable which has instance for
each thread and RAII to ensure that the appropriate flag is set to
`true` on entering commit section and reset to `false` after commit from
reanimated on UI thread.

## Test plan

- enable state reconciliation for commit from `reanimated` and check
whether the code goes into `isReanimatedCommit` `if` body in
`ReanimatedCommitHook`

or

- create a custom commit hook and register it before
`ReanimatedCommitHook` and perform same check as above

kowczarz committed:
Fix `useAnimatedSensor` vector velocity on Android (#4933)

## Summary

Velocity of gyroscope and magnetic field sensors were inverted when
compared to iOS.

## Test plan

Added repro in examples.

---------

Co-authored-by: Kamil Owczarz <kamil.owczarz@swmanison.com>
Co-authored-by: Tomek Zawadzki <tomekzawadzki98@gmail.com>

tomekzaw commented on src/reanimated2/js-reanimated/JSReanimated.ts:
```diff
@@ -101,54 +101,72 @@ export default class JSReanimated {
+      case SensorType.GRAVITY:
+        return () => {
+          let { x, y, z } = sensor;
+          [x, y, z] =
+            this.platform === Platform.WEB_ANDROID ? [-x, -y, -z] : [x, y, z];
```
I know that you didn't touch this code but could we unify it with the case below?
```suggestion

          // Android sensors have a different coordinate system than iOS
          if (this.platform === Platform.WEB_ANDROID) {
            [x, y, z] = [-x, -y, -z];
          }

```

tjzel committed:
Use original function's name inside worklet (#4951)

<!-- Thanks for submitting a pull request! We appreciate you spending
the time to work on these changes. Please follow the template so that
the reviewers can easily understand what the code changes affect. -->

## Summary

Currently plugin uses `_f` as private function name inside of a
worklet's function factory. However, @m-bert requested to use original
function names to use for layout animations on web.

I added funny scripts as a bonus.

## Test plan

Unit tests, CI and Example app run

bartlomiejbloniarz committed:
Add `reduceMotion` method to animation builders (#4884)

<!-- Thanks for submitting a pull request! We appreciate you spending
the time to work on these changes. Please follow the template so that
the reviewers can easily understand what the code changes affect. -->

## Summary
This PR extends the `BaseAnimationBuilder`, `Keyframe` and
`SharedTransition` with a `reduceMotion(config)` method. Possible values
of config are `ReduceMotion.System`, `ReduceMotion.Always` and
`ReduceMotion.Never`. It defaults to `system`.

In `BaseAnimationBuilder` and `Keyframe` the config is passed down to
animation functions.

Exiting animations and shared transitions are completely removed when
the config is set to `always` or it is set to `system` and the `Reduce
Motion` accessibility setting is enabled.
### LA example

https://github.com/software-mansion/react-native-reanimated/assets/56109050/5f618e78-83a4-4c47-ae3a-405be1cf24dc

### SET example

https://github.com/software-mansion/react-native-reanimated/assets/56109050/bfec2de3-cd6d-49cd-8efb-5ebd3d9508fa

(The disappearing text and other problems in this example are linked to
some issues with custom animations in SETs)

## Test plan

Test the behavior of the example app with the accessibility setting on
and off.

wojtus7 committed:
Unify the code

wojtus7 committed:
Merge remote-tracking branch 'mainstream/main' into wojtus7/web-android-sensor
